File: composeApp/src/androidMain/kotlin/app/xl/sportappkmp/exercisesTab/ExercisesScreen.kt
package app.xl.sportappkmp.exercisesTab

import androidx.compose.foundation.background
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.PaddingValues
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.statusBarsPadding
import androidx.compose.foundation.lazy.LazyColumn
import androidx.compose.foundation.lazy.itemsIndexed
import androidx.compose.material3.ExperimentalMaterial3Api
import androidx.compose.material3.HorizontalDivider
import androidx.compose.material3.ModalBottomSheet
import androidx.compose.material3.rememberModalBottomSheetState
import androidx.compose.runtime.Composable
import androidx.compose.runtime.getValue
import androidx.compose.ui.Modifier
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.platform.LocalContext
import androidx.compose.ui.unit.dp
import androidx.lifecycle.compose.collectAsStateWithLifecycle
import androidx.lifecycle.viewmodel.compose.viewModel
import app.xl.sportappkmp.MR
import app.xl.sportappkmp.exercisesTab.views.ExerciseListRow
import app.xl.sportappkmp.exercisesTab.views.FilterBar
import app.xl.sportappkmp.exercisesTab.views.SearchBar
import app.xl.sportappkmp.exercisesTab.views.EnumPickerSheet
import app.xl.sportappkmp.models.Equipment
import app.xl.sportappkmp.models.FilterSheetType
import app.xl.sportappkmp.models.MuscleGroup
import app.xl.sportappkmp.models.UnitType
import app.xl.sportappkmp.utils.Localizer
import app.xl.sportappkmp.viewModels.exercisesTab.ExercisesScreenViewModel

@OptIn(ExperimentalMaterial3Api::class)
@Composable
fun ExercisesScreen(
    modifier: Modifier = Modifier,
    viewModel: ExercisesScreenViewModel = viewModel(),
    onExerciseClick: (String) -> Unit
) {
    val state by viewModel.state.collectAsStateWithLifecycle()
    val activeSheet by viewModel.activeSheet.collectAsStateWithLifecycle()

    val sheetState = rememberModalBottomSheetState(
        skipPartiallyExpanded = true
    )

    val context = LocalContext.current
    val localizer = Localizer(context)


    Column(
        modifier = modifier
            .fillMaxSize()
            .statusBarsPadding()
            .background(Color(MR.colors.baseGray.getColor(context)))
    ) {
        SearchBar(
            modifier = Modifier.padding(horizontal = 24.dp),
            query = state.query,
            onQueryChange = { query ->
                viewModel.onQueryChange(query = query)
            },
            onResetQuery = {
                viewModel.resetQuery()
            }
        )

        Spacer(modifier = Modifier.height(16.dp))

        FilterBar(
            modifier = Modifier.padding(horizontal = 24.dp),
            showResetButton = state.equipment != null
                    || state.muscleGroup != null
                    || state.unitType != null,
            equipmentTitle = state.equipment?.localizedTitle(localizer)
                ?: localizer.get(MR.strings.allEquipment),
            muscleGroupTitle = state.muscleGroup?.localizedTitle(localizer)
                ?: localizer.get(MR.strings.allMuscles),
            unitTypeTitle = state.unitType?.localizedTitle(localizer)
                ?: localizer.get(MR.strings.allUnitTypes),
            onEquipmentClick = { viewModel.openSheet(FilterSheetType.EQUIPMENT) },
            onMuscleClick = { viewModel.openSheet(FilterSheetType.MUSCLE) },
            onUnitTypeClick = { viewModel.openSheet(FilterSheetType.UNIT_TYPE)},
            onResetFilterClick = { viewModel.resetFilters() }
        )

        Spacer(modifier = Modifier.height(16.dp))

        HorizontalDivider(
            modifier = Modifier
                .fillMaxWidth()
                .padding(horizontal = 24.dp),
            thickness = 1.dp,
            color = Color.DarkGray
        )

        LazyColumn(
            modifier = Modifier.fillMaxSize(),
            contentPadding = PaddingValues(horizontal = 24.dp)
        ) {
            itemsIndexed(
                items = state.exercises,
                key = { _, exercise -> exercise.id }
            ) { index, exercise ->
                ExerciseListRow(
                    localizer = localizer,
                    exercise = exercise,
                    onClick = onExerciseClick
                )
            }
        }
    }

    if (activeSheet != FilterSheetType.NONE) {
        ModalBottomSheet(
            sheetState = sheetState,
            onDismissRequest = { viewModel.closeSheet() },
            containerColor = Color.White
        ) {
            when (activeSheet) {
                FilterSheetType.EQUIPMENT -> EnumPickerSheet(
                    localizer = localizer,
                    items = Equipment.allCases,
                    onSelect = {
                        viewModel.onEquipmentChange(it)
                        viewModel.closeSheet()
                    }
                )

                FilterSheetType.MUSCLE -> EnumPickerSheet(
                    localizer = localizer,
                    items = MuscleGroup.allCases,
                    onSelect = {
                        viewModel.onMuscleGroupChange(it)
                        viewModel.closeSheet()
                    }
                )

                FilterSheetType.UNIT_TYPE -> EnumPickerSheet(
                    localizer = localizer,
                    items = UnitType.allCases,
                    onSelect = {
                        viewModel.onUnitTypeChange(it)
                        viewModel.closeSheet()
                    }
                )

                // TODO: - can't use the optional FilterSheetType - crash on iOS
                FilterSheetType.NONE -> Unit
            }
        }
    }
}

File: iosApp/iosApp/Presentation/ExercisesTab/ExercisesScreen.swift
//
//  ExercisesScreen.swift
//  iosApp
//
//  Created by Valeriy P on 28.12.2025.
//

import Shared
import SwiftUI

struct ExercisesScreen: View {

    @StateObject private var viewModel = ExercisesScreenViewModel()
    @State private var path = NavigationPath()

    private let localizer = Localizer()

    private var screenState: ExerciseListState {
        viewModel.state(
            \.state,
            equals: ==,
            mapper: { $0 }
        )
    }

    private var queryBinding: Binding<String> {
        .init(
            get: { screenState.query },
            set: viewModel.onQueryChange
        )
    }

    private var activeSheetBinding: Binding<FilterSheetType?> {
        Binding(
            get: {
                let sheet = viewModel.state(
                    \.activeSheet,
                    equals: { $0 == $1 },
                    mapper: { $0 }
                )
                return sheet == .none ? nil : sheet
            },
            set: { newValue in
                if let sheet = newValue {
                    viewModel.openSheet(type: sheet)
                } else {
                    viewModel.closeSheet()
                }
            }
        )
    }

    var equipmentTitle: String {
        localizer.getOpt(id: screenState.equipment?.titleRes)
            ?? localizer.get(id: MR.strings().allEquipment)
    }

    var muscleGroupTitle: String {
        localizer.getOpt(id: screenState.muscleGroup?.titleRes)
            ?? Localizer().get(id: MR.strings().allMuscles)
    }

    var unitTypeTitle: String {
        localizer.getOpt(id: screenState.unitType?.titleRes)
            ?? Localizer().get(id: MR.strings().allUnitTypes)
    }

    var showResetButton: Bool {
        screenState.equipment != nil
            || screenState.muscleGroup != nil
            || screenState.unitType != nil
    }

    var body: some View {
        NavigationStack(path: $path) {
            VStack(spacing: 0) {
                TextField(
                    localizer.get(id: MR.strings().searchBarPlaceholder),
                    text: queryBinding
                )
                .padding()
                .overlay {
                    RoundedRectangle(cornerRadius: 12)
                        .stroke(Color.gray, lineWidth: 1)

                    if !queryBinding.wrappedValue.isEmpty {
                        HStack {
                            Spacer()
                            Button {
                                queryBinding.wrappedValue = ""
                            } label: {
                                Image(systemName: "xmark.circle.fill")
                                    .foregroundColor(.gray)
                                    .padding(.trailing, 12)
                            }
                            .transition(.opacity.combined(with: .scale))
                        }
                    }
                }
                .background(.white)

                Spacer()
                    .frame(height: 16)

                ExercisesFilterBar(
                    equipmentTitle: equipmentTitle,
                    muscleGroupTitle: muscleGroupTitle,
                    unitTypeTitle: unitTypeTitle,
                    showResetButton: showResetButton,
                    onEquipmentTap: {
                        viewModel.openSheet(type: .equipment)
                    },
                    onMuscleGroupTap: {
                        viewModel.openSheet(type: .muscle)
                    },
                    onUnitTypeTap: {
                        viewModel.openSheet(type: .unitType)
                    },
                    onResetButtonTap: {
                        viewModel.resetFilters()
                    }
                )

                Spacer()
                    .frame(height: 16)

                Divider()

                ScrollView {
                    LazyVStack(alignment: .leading, spacing: 0) {
                        ForEach(screenState.exercises, id: \.id) { exercise in
                            NavigationLink(
                                value: ExercisesRoute.detail(exercise.id)
                            ) {
                                ExerciseListRow(exercise: exercise)
                            }
                            .buttonStyle(.plain)
                        }
                    }
                }
            }
            .navigationTitle(
                localizer.get(id: MR.strings().exercisesTabTitle)
            )
            .navigationBarTitleDisplayMode(.inline)
            .toolbar {
                ToolbarItem(placement: .navigationBarTrailing) {
                    Button(action: {
                        path.append(ExercisesRoute.create)
                    }) {
                        Image(systemName: "plus")
                    }
                }
            }
            .toolbarBackground(
                Color(MR.colors().baseGray.getUIColor()),
                for: .navigationBar
            )
            .toolbarBackground(.visible, for: .navigationBar)
            .navigationDestination(for: ExercisesRoute.self) { route in
                destination(for: route)
            }
            .sheet(item: activeSheetBinding) { sheet in
                getSheet(for: sheet)
            }
            .padding(.horizontal, 20)
            .background(Color(MR.colors().baseGray.getUIColor()))
        }
    }
    
    @ViewBuilder
    private func destination(for route: ExercisesRoute) -> some View {
        switch route {
        case .detail(let id):
            ExercisesInfoScreen(path: $path, exerciseId: id)
        case .edit(let id):
            ExerciseCreateScreen(id: id)
        case .create:
            ExerciseCreateScreen(id: nil)
        }
    }

    @ViewBuilder
    private func getSheet(for sheet: FilterSheetType) -> some View {
        switch sheet {
        case .equipment:
            EnumPickerSheet(
                items: Equipment.companion.allCases,
                onSelect: { selected in
                    viewModel.onEquipmentChange(equipment: selected)
                    viewModel.closeSheet()
                }
            )

        case .muscle:
            EnumPickerSheet(
                items: MuscleGroup.companion.allCases,
                onSelect: { selected in
                    viewModel.onMuscleGroupChange(muscleGroup: selected)
                    viewModel.closeSheet()
                }
            )

        case .unitType:
            EnumPickerSheet(
                items: UnitType.companion.allCases,
                onSelect: { selected in
                    viewModel.onUnitTypeChange(unitType: selected)
                    viewModel.closeSheet()
                }
            )

        default:
            EmptyView()
        }
    }
}

#Preview {
    ExercisesScreen()
}
